src/main/scala/org/broadinstitute/dsde/workbench/sam/schema/JndiSchemaDAO.scala

```diff
@@ -49,40 +52,190 @@ object JndiSchemaDAO {
   }
 }
 
-class JndiSchemaDAO(protected val directoryConfig: DirectoryConfig)(implicit executionContext: ExecutionContext) extends JndiSupport with DirectorySubjectNameSupport {
+object SchemaStatus {
+  sealed trait SchemaStatus
+
+  case object Wait extends SchemaStatus
+  case object Proceed extends SchemaStatus
+  case object Ignore extends SchemaStatus
+}
+
+class JndiSchemaDAO(protected val directoryConfig: DirectoryConfig, val schemaLockConfig: SchemaLockConfig)(implicit executionContext: ExecutionContext) extends JndiSupport with DirectorySubjectNameSupport with LazyLogging {
 
   def init(): Future[Unit] = {
+    if(schemaLockConfig.lockSchemaOnBoot) { initWithSchemaLock() }
+    else { createAndDestroySchema().map(_ => ()) }
+  }
+
+  private def initWithSchemaLock(): Future[Unit] = {
+    for {
+      schemaLockStatus <- initSchemaLock()
+      schemaUpdateStatus <- schemaLockStatus match {
+        case Proceed => createAndDestroySchema()
+        case Ignore => Future.successful(Ignore)
+      }
+      _ <- schemaUpdateStatus match {
+        case Proceed => setSchemaUpdateComplete()
+        case Ignore => Future.successful(Ignore)
+      }
+    } yield ()
+  }
+
+  private def initSchemaLock(): Future[SchemaStatus] = {
+    logger.info("Acquiring schema lock...")
+
+    for {
+      _ <- createSchemaChangeLockSchema()
+      _ <- createOrgUnit(schemaLockOu)
+      result <- lockSchema()
+    } yield result
+  }
+
+  private def createAndDestroySchema(): Future[SchemaStatus] = {
     for {
       _ <- destroySchema()
       _ <- createSchema()
-    } yield ()
+    } yield Proceed
+  }
+
+  private def setSchemaUpdateComplete(): Future[Unit] = withContext { ctx =>
+    val myAttrs = new BasicAttributes(true)
+    myAttrs.put("completed", true.toString)
+
+    logger.info(s"Schema successfully updated to version [${schemaLockConfig.schemaVersion}]")
+
+    ctx.modifyAttributes(schemaLockDn(schemaLockConfig.schemaVersion), DirContext.REPLACE_ATTRIBUTE, myAttrs)
+  }
+
+  private def readSchemaStatus(): Future[SchemaStatus] = withContext { ctx =>
+    val attributes = Try { ctx.getAttributes(schemaLockDn(schemaLockConfig.schemaVersion)) }
```

Comment:
A note of caution: Since the locks are schema version specific, all sam instances must be configured for the same schema version. Otherwise they will be locking on different objects and not blocking each other.

The only time this would be a problem is if 2 instances are using 2 different versions that both don't exist which is probably extremely unlikely.